internal/rbd/clone.go

```diff
@@ -0,0 +1,245 @@
+		}
+		err = tempClone.deleteSnapshot(ctx, snap)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to delete snapshot: %v"), err)
+			return false, err
+		}
+		return true, nil
+	}
+	// as the temp clone doesnot exists,check snapshot exists on parent volume
+	err = parentVol.checkSnapExists(snap)
+	if err == nil {
+		// the temp clone exists, delete it lets reserve a new ID and
+		// create new resources for clearner approach
+		err = parentVol.deleteSnapshot(ctx, snap)
+	}
+	if errors.As(err, &esnf) {
+		return false, nil
+	}
+	return false, err
+}
+
+func (rv *rbdVolume) generateTempClone() *rbdVolume {
+	tempClone := rbdVolume{}
+	tempClone.conn = rv.conn
+	// The temp clone image need to have deep flatten feature
+	f := []string{librbd.FeatureNameLayering, librbd.FeatureNameDeepFlatten}
+	tempClone.imageFeatureSet = librbd.FeatureSetFromNames(f)
+	tempClone.ClusterID = rv.ClusterID
+	tempClone.Monitors = rv.Monitors
+	tempClone.Pool = rv.Pool
+	// The temp cloned image name will be always (rbd image name + "-temp")
+	// this name will be always unique, as cephcsi never creates an image with
+	// this format for new rbd images
+	tempClone.RbdImageName = rv.RbdImageName + "-temp"
+	return &tempClone
+}
+
+func (rv *rbdVolume) createCloneFromImage(ctx context.Context, parentVol *rbdVolume) error {
+	// generate temp cloned volume
+	tempClone := rv.generateTempClone()
+	snap := &rbdSnapshot{
+		RbdSnapName: rv.RbdImageName,
+		Pool:        rv.Pool,
+	}
+
+	var (
+		errClone   error
+		errFlatten error
+		err        error
+	)
+	var efip ErrFlattenInProgress
+	var j = &journal.Connection{}
+
+	j, err = volJournal.Connect(rv.Monitors, rv.conn.Creds)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+	defer j.Destroy()
+
+	// create snapshot and temporary clone and delete snapshot
+	err = createRBDClone(ctx, parentVol, tempClone, snap, rv.conn.Creds)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err != nil || errFlatten != nil {
+			if !errors.Is(errFlatten, &efip) {
+				// cleanup snapshot
+				cErr := cleanUpSnapshot(ctx, parentVol, snap, tempClone, rv.conn.Creds)
+				if cErr != nil {
+					klog.Errorf(util.Log(ctx, "failed to cleanup image %s or snapshot %s: %v"), snap, tempClone, cErr)
+				}
+			}
+		}
+		if err != nil || errClone != nil {
+			cErr := cleanUpSnapshot(ctx, tempClone, snap, rv, rv.conn.Creds)
+			if cErr != nil {
+				klog.Errorf(util.Log(ctx, "failed to cleanup image %s or snapshot %s: %v"), snap, tempClone, cErr)
+			}
+		}
+	}()
+	// flatten clone
+	errFlatten = tempClone.flattenRbdImage(ctx, rv.conn.Creds, false, rbdHardMaxCloneDepth, rbdSoftMaxCloneDepth)
+	if errFlatten != nil {
+		return err
+	}
+	// create snap of temp clone from temporary cloned image
+	// create final clone
+	// delete snap of temp clone
+	errClone = createRBDClone(ctx, tempClone, rv, snap, rv.conn.Creds)
+	if errClone != nil {
+		// set iErr error to cleanup temporary snapshot and temporary clone
```

Comment:
reference to the now renamed `iErr`

**Many thanks for all the additional comments and better naming of variables!**